include/engine/routing_algorithms/routing_base.hpp

```diff
@@ -41,41 +41,23 @@ bool needsLoopForward(const PhantomNode &source_phantom, const PhantomNode &targ
 
 bool needsLoopBackwards(const PhantomNode &source_phantom, const PhantomNode &target_phantom);
 
-template <bool DIRECTION, typename Heap>
-void insertNodesInHeap(Heap &heap, const PhantomNode &phantom_node)
-{
-    BOOST_ASSERT(phantom_node.IsValid());
-
-    const auto weight_sign = DIRECTION == FORWARD_DIRECTION ? -1 : 1;
-    if (phantom_node.forward_segment_id.enabled)
-    {
-        heap.Insert(phantom_node.forward_segment_id.id,
-                    weight_sign * phantom_node.GetForwardWeightPlusOffset(),
-                    phantom_node.forward_segment_id.id);
-    }
-    if (phantom_node.reverse_segment_id.enabled)
-    {
-        heap.Insert(phantom_node.reverse_segment_id.id,
-                    weight_sign * phantom_node.GetReverseWeightPlusOffset(),
-                    phantom_node.reverse_segment_id.id);
-    }
-}
-
 template <bool DIRECTION>
 void insertNodesInHeap(SearchEngineData<ch::Algorithm>::ManyToManyQueryHeap &heap,
                        const PhantomNode &phantom_node)
 {
     BOOST_ASSERT(phantom_node.IsValid());
 
     const auto weight_sign = DIRECTION == FORWARD_DIRECTION ? -1 : 1;
-    if (phantom_node.forward_segment_id.enabled)
+    if ((DIRECTION == FORWARD_DIRECTION && phantom_node.IsForwardValidSource()) ||
```

Comment:
`DIRECTION` is a little bit of a mis-normer here. In the `ManyToMany` plugin we insert each source and target in forward and backward direction.

It would make more sense to have two functions `insertSourceInHeap` and `insertTargetInHeap`.